### PR TITLE
Helm Chart + bootstrap installation script for Harbor

### PR DESCRIPTION
**Pull request by nuzz**
A bootstrap script + Helm Chart to install a fully functioning Harbor registry on a Kubernetes cluster. This chart is intended to be provider agnostic. This installation has been tested on AWS (self managed k8s cluster), Google GKE, and Micorsoft AKS. 

See README.md for more information

### CODE REVIEW

**Comment by reasonerjt** [comment]
Thanks for the PR! 
We're going to review this but please bear with the delay as there are other focus.

File: contrib/helm-chart/templates/nginx-svc.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: nginx
+  annotations:
+    {{- range $key, $value := .Values.nginx.annotations }}
+      {{ $key }}: {{ $value | quote }}
+    {{- end }}
+spec:
+  type: LoadBalancer
```
Comment (xeor) [inline contrib/helm-chart/templates/nginx-svc.yaml]: Can we get a NodePort alternative as well? Maybe something like https://github.com/kubernetes/charts/blob/master/stable/postgresql/templates/svc.yaml?

File: contrib/helm-chart/templates/postgres-statefulset.yaml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            cpu: 100m
+            memory: 512Mi
+        volumeMounts:
+        - name: pgdata
+          mountPath: /var/lib/postgresql
+        ports:
+        - containerPort: 5432
+          name: postgres-port
+          protocol: TCP
+  volumeClaimTemplates:
```
Comment (xeor) [inline contrib/helm-chart/templates/postgres-statefulset.yaml]: Need to have storageClassName configurable. Defining that in annotations is being deprecated

File: contrib/helm-chart/templates/registry-cm.yaml [hunk 3]
```diff
@@ -0,0 +1,51 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: registry-config
+data:
+  config.yml: |+
+    version: 0.1
+    log:
+      level: info
```
Comment (xeor) [inline contrib/helm-chart/templates/registry-cm.yaml]: configurable log-level?

**Comment by xeor** [comment]
Looking forward for this PR :) But can you also implement:

* Configurable storageclassnames
* Configurable namespace (`{{ .Release.Namespace }}`)
* Don't rely on people having `LoadBalancer` available..
* I'm not sure rbac is in scope.. But a nice project to generate rbac is https://github.com/liggitt/audit2rbac..
* Same with https://github.com/kubernetes/charts/blob/master/stable/postgresql/templates/networkpolicy.yaml

Some might be for a `v2` :)

**Comment by paulczar** [comment]
@nuzz I've also been working on helm charts for harbor,  it looks like we've done fairly similar stuff but we've each made a bit more progress in different ways - https://github.com/paulczar/helm-harbor.

**Comment by paulczar** [comment]
I'm not a big fan of the bootstrap scripts, I'd rather see defaults that make it work out of the box for minikube installations and documentation on what to change for different install environments.

I'm also not a huge fan of the seperate secrets file loading into a common namespace, I'd rather see the secrets stick to being part of the same namespace as the harbor install.  I would also rather keep the secrets as values set in helm ( I know helm doesn't fully support proper secrets, but its coming real soon https://github.com/kubernetes/helm/pull/2721 ).

**Comment by paulczar** [comment]
I've been working with the upstream helm charts team with my own harbor charts here - https://github.com/kubernetes/charts/pull/3383

**Comment by jessehu** [comment]
Hello guys, I'm Jesse Hu from Harbor dev team, will in charge of making Harbor Helm Chart as a formal community feature into harbor github repo. Thanks very much to @paulczar @nuzz @xeor for sharing your harbor chart code, and the PR https://github.com/kubernetes/charts/pull/3383 for adding incubator/harbor chart.
I think the patch by @paulczar https://github.com/paulczar/helm-harbor is a good start, so could you @paulczar please send a formal pull request :)
The bootstrap script in @nuzz is also a good idea which is used to generate random password and SSL certificates. And I think we can consider using the chart template function to implement the bootstrap script https://masterminds.github.io/sprig/crypto.html, then it's a part of the harbor chart, not a standalone script outside harbor chart.
Any thought? :)

**Comment by jessehu** [comment]
@paulczar I have a fork to make your code work with latest Harbor 1.4.0 release  https://github.com/jessehu/helm-harbor/commit/c2818c77576b799b84fa6d78685d44fa9c6588b4

**Comment by jessehu** [comment]
We have merged https://github.com/vmware/harbor/pull/4271

**Comment by reasonerjt** [comment]
Thanks, @paulczar, @jessehu and @nuzz 
I'm closing this PR.

**Comment by jessehu** [comment]
Hi @nuzz @xeor and other guys interested in Harbor Helm Chart, we have merged  https://github.com/vmware/harbor/pull/4373 'Update Harbor helm chart to deploy Harbor 1.4.0 release' into harbor master branch https://github.com/vmware/harbor/tree/master/contrib/helm/harbor.

Welcome to try it out and your continuous idea and code contribution to Harbor are still appreciated. Thanks!

**Pull request closed without merging**